### PR TITLE
Allow interpolated string adjacent to '=' (e.g. C(Name=$"value"))

### LINKED ISSUE #16696: Property initialization doesn't work without space before interpolated string
**Repro steps**
```fsharp
type C() =
    member val Name = "" with get, set
    
let x = C(Name="123") //works
let y = C(Name=$"123") // doesn't work
let z = C(Name= $"123") // works
```
[Sharplab](https://sharplab.io/#v2:DYLgZgzgPgLgngBwKYAIDCAKAlCgvAWACgUSUBbJMgIyQCcUA3AQ2BQDkmK8UAiHlAO4BLGAAsUAcyQwANCgjSipFEWDSUAD26YOFXDwCMAJgDMPHAHoLAgPa0A1hFXq42jLqS4AJIdPmUVigAJjZIEAB2AOQwgnb2QA)

**Expected behavior**

All three cases should work

**Actual behavior**

Second case doesn't work

**Known workarounds**

Add space before value

**Related information**

.NET 8
F# 8

### PR DESCRIPTION
**Pull request by edgarfgp**
Fixes [#16696](https://github.com/dotnet/fsharp/issues/16696).

## Problem

When an interpolated string immediately follows `=` with no space, the lexer greedily matches `=$` as a single `INFIX_COMPARE_OP` token (since `$` is an operator character). `checkExprOp` then rejects it with FS0035 (`'$' is not permitted as a character in operator names`), so property/named-argument initialization with an interpolated string fails to parse.

## Before

```fsharp
type C() =
    member val Name = "" with get, set

let a = C(Name="123")     // works
let b = C(Name=$"123")    // error FS0035: '$' is not permitted as a character in operator names
let c = C(Name= $"123")   // works (space before $)
```

```fsharp
let x =$"123"             // error FS0035
let y =$"hello {world}"   // error FS0035
```

## After

```fsharp
let b = C(Name=$"123")            // works → "123"
let x =$"hello {world}"           // works → "hello world"
let y =$"%d{n}"                   // works → typed hole
let z = { Name=$"value" }         // works in record creation
let w = { r with Name=$"value" }  // works in record copy
let t =$"""triple {x}"""          // works for triple-quoted
```

The produced AST is identical to the spaced form `= $"..."`, so all downstream tooling sees the same tree.

## Solution

A lexer-level rule for the exact 3-character sequence `=$"`:

```fsharp
| '=' '$' '"' {
      lexbuf.LexemeLength <- 1
      lexbuf.EndPos <- lexbuf.StartPos.ShiftColumnBy(1)
      EQUALS }
```

It matches `=$"` (winning over the 2-char operator rule via longest-match), emits `EQUALS`, then **rewinds** `LexemeLength` to 1 so the next scan begins at `$"`. The existing interpolated-string lexer then processes the rest — including interpolation holes and triple-quoted forms. Because the fix is in the lexer, it applies in every position where `= $"..."` is valid (let-bindings, named args, record creation/copy), not just one grammar rule.

The previously-`internal` `LexBuffer.LexemeLength` setter is exposed in `prim-lexing.fsi` (no change to the public API surface — the type is `internal`).

## Scope

Limited to the `$"` opening sequence (single- and triple-quoted). Defining or using `=$` as a custom operator remains rejected (FS0035), preserving the "reserved for future use" semantics.

## Open question — should we also handle the remaining variants?

These forms still require a space and continue to error as before (regression tests assert this):

```fsharp
let x =$@"abc"        // verbatim interpolated   ($@)
let y =@$"abc"        // verbatim interpolated   (@$)
let z =$$"""abc"""    // extended interpolated   ($$)
```

They're out of scope for the reported issue (which only shows the `$"` form), and the workaround is a single space. They could be added with the same lexer technique (longer `=$@"`, `=@$"`, `=$$"""` rules with a larger rewind), at the cost of more lexer rules. **Do we want to cover them in this PR, defer to a follow-up, or leave them as-is?**

### CODE REVIEW

**Review by T-Gro — APPROVED** [review]
Clean, minimal fix at exactly the right layer. The approach is sound:

**Correctness**
- The 3-char rule '=' '$' '"' wins by longest-match over the 2-char =$ that the INFIX_COMPARE_OP rule would match (since " is not in op_char). This is the key insight that makes the fix safe.
- The rewind via LexemeLength <- 1 + EndPos adjustment correctly lets the regular interpolated-string lexer take over at $", including triple-quoted forms ($"""...""" — first three chars =$" still trigger the rule, then next scan sees $""" and the existing triple-quote handling kicks in).
- Negative cases (=, defining (=$), infix =$) all still route through checkExprOp and reject correctly.

**API surface**
- LexemeLength setter is exposed in prim-lexing.fsi but LexBuffer<'Char> is 	ype internal — no public API impact.

**CI**: All 50 check runs green.

**Tests**: Excellent coverage — positive cases (let-bindings, named args, records, record-copy, holes, typed holes), negative cases (non-quote, operator definition, infix usage, verbatim out-of-scope), syntax-tree baselines, and the space-separated form unchanged.

**One minor suggestion** (non-blocking): consider adding a component test that compiles and runs =$"""triple {x}""" alongside the syntax-tree baseline. The baseline proves parsing but not codegen/execution for that form.

Nice work!

File: tests/FSharp.Compiler.ComponentTests/Language/InterpolatedStringsTests.fs [hunk 1]
```diff
@@ -361,3 +361,107 @@ let s = $"{f.Invoke(42)}"
...
+    let ``Issue 16696 - let-binding with =$"..." (no space)`` () =
+        Fsx """
+let n = 42
+let world = "world"
+let plain =$"123"
+let withHole =$"hello {world}"
+let withTypedHole =$"%d{n}"
+if plain <> "123" then failwithf "plain: expected 123, got %s" plain
+if withHole <> "hello world" then failwithf "withHole: expected 'hello world', got %s" withHole
+if withTypedHole <> "42" then failwithf "withTypedHole: expected 42, got %s" withTypedHole
```
Comment (auduchinok) [inline tests/FSharp.Compiler.ComponentTests/Language/InterpolatedStringsTests.fs]: @edgarfgp Could you minimize the test, so it's clear what it tests?

File: src/Compiler/lex.fsl [hunk 2]
```diff
@@ -975,8 +975,17 @@ rule token (args: LexArgs) (skip: bool) = parse
...
  | ignored_op_char* ('@'|'^')                  op_char* { checkExprOp lexbuf; INFIX_AT_HAT_OP(lexeme lexbuf) }
 
+ // For '=$"' (property/named-arg initialization with an interpolated string, e.g. C(Name=$"123")):
+ // match the 3 chars, but consume only the '=' and rewind so the next scan begins at '$"',
+ // letting the regular interpolated-string lexer process it (including any '{...}' holes).
+ // See https://github.com/dotnet/fsharp/issues/16696.
+ | '=' '$' '"' {
+       lexbuf.LexemeLength <- 1
+       lexbuf.EndPos <- lexbuf.StartPos.ShiftColumnBy(1)
+       EQUALS }
```
Comment (auduchinok) [inline src/Compiler/lex.fsl:985]: @edgarfgp What other implications may this change bring? Have you considered following approach used for operators and type arguments?

**Review by auduchinok — CHANGES_REQUESTED** [review]
This change feels a bit too ad-hoc. It workarounds a specific case arising due to non-idiomatic formatting of the code, i.e. due to the missing spaces around `=`.

@edgarfgp Since this PR changes what is allowed by the language, have you considered changing the rule that causes the problem in the first place instead of trying to workaround it?

The workaround is done by exposing more lexer internals and doing processing after a token is lexed. Modifying tokens like this is sometimes done in the compiler already, but for operators it's done in LexFilter. Have you compared your approach with the existing one?

The tests look as if they are AI-generated: there's a lot of extra code that hides the parts being tested, and the tests effectively duplicate each other, since they only test simple adjacent `=` and `$`, while not testing how different kinds of interpolations interoperate (e.g. normal string interpolations inside triple quote ones), or whether operators are lexed differently due to this change. I think we should deduplicate the syntax tests and add more cases highlighting the lexing.

File: src/Compiler/lex.fsl (hunk 2 above)
Comment (edgarfgp) [reply src/Compiler/lex.fsl:985]: @auduchinok Base on my investigation

> What other implications may this change bring?

This rule is higher priority for the exact sequence `=$"`. So it can only win when the input literally starts with `=$"` .

The only thing that changes is that immediately before the `$` now lexes as EQUALS + interpolated string. The token `=$` before a quote as never valid anyway. Everything else is identical, every `$` operator is still invalid

> Have you considered following approach used for operators and type arguments?

Yes, I tried first but the interpolation has to be made while scanning the string characters which only happens in the lexer. LexFilter runs after that which is too late.

Comment (auduchinok) [reply src/Compiler/lex.fsl:985]: Thanks @edgarfgp!

File: tests/FSharp.Compiler.ComponentTests/Language/InterpolatedStringsTests.fs (hunk 1 above)
Comment (edgarfgp) [reply tests/FSharp.Compiler.ComponentTests/Language/InterpolatedStringsTests.fs]: Done

**Comment by edgarfgp** [comment]
@auduchinok Are you satisfied with the fix rationale or you have a better way of fixing this ?.

**Review by auduchinok — APPROVED** [review]

**Comment by edgarfgp** [comment]
@T-Gro This is ready!. Will raise follow up PRs with the remaining open questions scenarios

**Review by T-Gro — APPROVED** [review]
